#  Computational Microbial Genomics: Peri-Implant Tissue Analysis
Repository for the project of Computational and Microbial Genomics (M.Sc. in Quantitative and Computational Biology, 2023-2024).

Group members: 
- Baldinelli Sara 
- De Pietri Letizia 
- Spadazzi Roan 

## Overview
This repository contains the computational analysis pipeline and results for a study on microbial genomes sampled from peri-implant tissues. The study focuses on understanding the taxonomy, genome annotation, pangenome characteristics, and phylogenetic relationships of microbial communities associated with different peri-implant conditions.

## Project Objectives
- Taxonomic Assignment: Classifying microbial genomes to understand their diversity and distribution in peri-implant tissues.
- Genome Annotation: Identifying coding sequences (CDSs) and hypothetical proteins (HPs) in the genomes.
- Pangenome Analysis: Exploring the core and accessory genes to assess whether the microbial community has an open or closed pangenome.
- Phylogenetic Analysis: Constructing phylogenetic trees to study the evolutionary relationships among microbial strains.


## Data and Methods
Data: 30 Metagenome-Assembled Genomes (MAGs) from peri-implant tissue samples categorized into healthy, mucositis, and peri-implantitis conditions.\
Tools:
- Quality Checking: CheckM
- Taxonomic Assignment: PhyloPhlAn
- Genome Annotation: Prokka
- Pangenome Analysis: Roary
- Phylogenetic Analysis: FastTree, MAFFT

## Key Findings
The SGB was assigned to the Metamycoplasmataceae family.\
The genome analysis suggests an open pangenome for the microbial community.\
Phylogenetic analysis revealed some clustering by condition, although no clear distinction was observed across all samples.
